keps/sig-storage/20190120-execution-hook-design.md

```diff
@@ -0,0 +1,520 @@
+
+The Create event of ExecutionHook will trigger the hook controller to run PodAction command. The status will be updated once it is available by controller. It is the caller's (user or controller) responsibility to delete the hook once the execution finishes (either succeeded or failed). Otherwise, ExecutionHook controller will garbage collect them after a certain amount of time.
+
+Here is the definition of an ExecutionHook API object:
+
+```
+// ExecutionHook is in the tenant namespace
+type ExecutionHook struct {
+        metav1.TypeMeta
+        // +optional
+        metav1.ObjectMeta
+
+        // Spec defines the behavior of a hook.
+        // +optional
+        Spec ExecutionHookSpec
+
+        // Status defines the current state of a hook.
+        // +optional
+        Status ExecutionHookStatus
+}
+```
+
+Here is the definition of the ExecutionHookSpec:
+
+```
+// PodActionName is copied to ExecutionHookSpec by the controller such as
+// the Snapshot Controller.
+type ExecutionHookSpec struct {
+        // PodSelection defines how to select pods and containers to run
+	// the executionhook. This field is required.
+        PodSelection podSelection
+
+        // Name of the PodAction. This is required.
+        PodActionName string
+}
+
+// PodSelection contains two fields, PodContainerNamesList and PodContainerSelector,
+// where one of them must be defined so that the hook controller knows where to
+// run the hook.
+type PodSelection struct {
+        // PodContainerNamesList lists the pods/containers on which the ExecutionHook
+        // should be executed. If not specified, the ExecutionHook controller will find
+        // all pods and containers based on PodContainerSelector.
+        // If both PodContainerNamesList and PodContainerSelector are not
+        // specified, the ExecutionHook cannot be executed and it will fail.
+        // +optional
+        PodContainerNamesList []PodContainerNames
+
+        // PodContainerSelector is for hook controller to find pods and containers
+        // based on the pod label selector and container names
+        // If PodContainerNamesList is specified, this field will not be used.
+        // +optional
+        PodContainerSelector *podContaienrSelector
+}
+
+type PodContainerNames struct {
+        PodName string
+        ContainerNames []string
+}
+
+type PodContainerSelector struct {
+	// PodSelector specifies a label query over a set of pods.
+        // +optional
+	PodSelector *metav1.LabelSelector
+
+        // ContainerSelector is a selector which must be true for the hook to run on the container
+        // Hook controller will find all containers in the selected pods based on ContainerSelector,
+        // if specified.
+        // The ContainerSeletor maps pod label to container name(s) so that the controller only select
```

Comment:
This is a bit hard to follow. What if a user wants to run the hook in a specific container in every pods selected by `PodSelector`? Can't this just be a list of container names, like how containers are selected in other Kubnernetes APIs?

Reply:
I think you are right, just a list is enough

Reply:
The problem with `container` in `HookAction` is users may use different container names even for the same type of containers, e.g., a mysql container running a vanilla mysql image. This prevents `HookAction` objects from being shared, which is the main point of making `HookAction` a separate resource kind.